cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandler.java

```diff
@@ -288,4 +292,17 @@ private ProgramType getProgramType(String programType) throws BadRequestExceptio
       throw new BadRequestException(String.format("Invalid program type '%s'", programType), e);
     }
   }
+
+  /**
+   *
+   * @param namespaceId namespace ID
+   * @param appId application ID
+   * @return latest app version
+   */
+  private String getLatestAppVersion(String namespaceId, String appId) {
+    ApplicationMeta latestApplicationMeta = store.getLatest(new NamespaceId(namespaceId), appId);
+    return latestApplicationMeta == null
+      ? ApplicationId.DEFAULT_VERSION
```

Comment:
so this DEFAULT_VERSION is to deal with the existing apps? Is there any other use case?

Reply:
yeah it's dealing with existing ones, also the returned version cannot be null

Reply:
why would it be null at the first place? It can only be null when there's no application right, then there's no point getting the default version

Reply:
it's correct. But unit tests are covering the no application case, which is checking the `-SNAPSHOT` version

Reply:
I agree with Sumeng - Latest will be null when -SNAPSHOT doesn't exist (when the app itself doesn't exist)

Reply:
getLatest returns null only when app doesn't exist - for existing apps we just scan the table to return any existing rows https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java#L361

Reply:
@sanjanasandeep yeah but you have to pass in a non-null version ID, otherwise the code will throw exception (validation on ApplicationID, or deeper exception from table)

Reply:
offline dicussed; we should return 404 when no latest version found

Reply:
I made a similar change, you basically need to add 
```
if (version == null) {
      throw new ApplicationNotFoundException(new ApplicationId(namespaceId.getNamespace(), appName));
    }
```

Reply:
updated